portability-worker/src/main/java/org/dataportabilityproject/worker/PortabilityInMemoryTransferCopier.java

```diff
@@ -37,95 +37,99 @@
-   * @param importAuthData The auth data for the import
-   * @param exportInformation Any pagination or resource information to use for subsequent calls.
-   */
-  private void copy(
-      Exporter exporter,
-      Importer importer,
-      AuthData exportAuthData,
-      AuthData importAuthData,
-      ExportInformation exportInformation) throws IOException {
-
-    logger.debug("copy iteration: {}", COPY_ITERATION_COUNTER.incrementAndGet());
-
-    // NOTE: order is important below, do the import of all the items, then do continuation
-    // then do sub resources, this ensures all parents are populated before children get
-    // processed.
-    logger.debug("Starting export, ExportInformation: {}", exportInformation);
-    ExportResult<?> exportResult = exporter.export(exportAuthData, exportInformation);
-    logger.debug("Finished export, results: {}", exportResult);
-
-    logger.debug("Starting import");
-    // TODO, use job Id?
-    importer.importItem("1", importAuthData, exportResult.getExportedData());
-    logger.debug("Finished import");
-
-    ContinuationData continuationData = (ContinuationData) exportResult.getContinuationData();
-
-    if (null != continuationData) {
-      // Process the next page of items for the resource
-      if (null != continuationData.getPaginationData()) {
-        logger.debug("start off a new copy iteration with pagination info");
-        copy(exporter, importer, exportAuthData, importAuthData,
-            new ExportInformation(continuationData.getPaginationData(),
-                exportInformation.getContainerResource()));
-      }
-
-      // Start processing sub-resources
-      if (continuationData.getContainerResources() != null && !continuationData
-          .getContainerResources().isEmpty()) {
-        for (ContainerResource resource : continuationData.getContainerResources()) {
-          copy(exporter, importer, exportAuthData, importAuthData,
-              new ExportInformation(null, resource));
-        }
-      }
+    private static final AtomicInteger COPY_ITERATION_COUNTER = new AtomicInteger();
+    private static final Logger logger = LoggerFactory
+            .getLogger(PortabilityInMemoryTransferCopier.class);
+
+    private final TransferServiceProviderRegistry registry;
+
+    public PortabilityInMemoryTransferCopier(TransferServiceProviderRegistry registry) {
+        this.registry = registry;
+    }
+
+    /**
+     * Kicks off the transfer job for the datatype from the export service to the import service.
+     *
+     * @param dataType       The data type to transfer
+     * @param exportService  The export service to transfer from
+     * @param exportAuthData The auth data for the export service
+     * @param importService  The import service to transfer to
+     * @param importAuthData The auth data for the import service
+     * @param jobId          The job id representing this copy
+     */
+    @Override
+    public void copyDataType(String dataType,
```

Comment:
A future refactor to consider, but I think would be difficult in the current intermediate state with some copying still being done in the API etc.

Each worker instance should be spun up with one (eventual) definition of all these properties (export service, auth data, job ID, etc) although they will not be available until a job is polled. WDYT of injecting all these fields with providers in the copier, so they are only evaluated once we start the copy, at which point the job is polled?

I think we should follow the same model for local worker testing too - spin up/tear down for each job tested locally

Reply:
Yes, we can refactor as to how to enforce the singleton nature of the data we want to pass around. Maybe an singleton annotated object with these params or individual data.